general/v0.1/protocol_basics.md

```diff
@@ -37,40 +36,93 @@ Communication between fediverse server and FASP MUST use HTTPS in production
+    "name": "fediexample",
+    "version": "6.2.7"
+  },
+  "protocols": [
+    "activitypub"
+  ],
+  "services": {
+    "outbound": [],
+    "inbound": []
+  },
+  "openRegistrations": false,
+  "metadata": {
+    "nodeName": "fedi",
+    "faspBaseUrl": "https://fedi.example.com/fasp"
+  }
+}
+```
+
+### Request Integrity
+
+In order to allow both parties to verify the integrity of message
+contents, all requests MUST contain a `content-digest` HTTP header as
+defined by [RFC-9530](https://tools.ietf.org/html/rfc9530.html).
+
+The hashing algorith used MUST be SHA-256.
+
+### Authentication
 
 As described in [03: Registration](registration.md) both FASP and
-fediverse server use OAuth 2.0 to authorize API calls. Both MUST obtain a valid
-access token and send this as a "bearer token" in the `Authorization`
-HTTP header with every API call.
+fediverse server exchange client IDs and public keys. API requests are
+being authenticated by HTTP Message Signatures as defined in
+[RFC-9421](https://tools.ietf.org/html/rfc9421.html).
 
-Example header:
+The signature algorithm used is EdDSA Using Curve edwards25519.
+Signatures cover signature parameters, the derived components `@method`
+and `@target-uri` and the HTTP header `content-digest` (see previous
+section).
+
+The `keyid` parameter MUST include the client ID and the corresponding 
+private key is used to generate the signature.
+
+The required signature parameters are `created` and `keyid`.
+
+Example headers:
 
 ```http
-Authorization: Bearer SpBr6rheOp891mwWOfT6Pb"
+Content-Digest: sha-256=:RK/0qy18MlBSVnWgjwz6lZEWjP/lF5HF9bvEF8FabDg=:
+Signature-Input: sig1=("@method" "@target-uri" "content-digest"); created=1728467285;
+keyid="b2ks6vm8p23w"
+Signature: sig1=:+CcncFjyE+JAuwJO8MOEhRdyfShQz59e9bWDYGN3hoBorVp69k4V2PvS7zJiAoX3QchMlc47sUF4DsptUN+rDQ==:
 ```
 
-Both FASP and fediverse server MUST expire access tokens, forcing the other
-side to periodically request a new one.
-
-The OAuth 2.0 endpoint to request an access token MUST reside at the
-path `/oauth/token` that is relative to the base URL as described
-above. Existing fediverse software that already uses
-OAuth 2.0 and wants to add FASP support cannot re-use existing
-routes. This simplifies FASP implementation and
-enables fediverse software implementers to separate their existing OAuth
-2.0 implementation for regular API clients from the FASP API if so
-desired.
+The signature MUST be verified by the receiving party using the public
+key belonging to the transmitted client ID (`keyid` parameter).
+
+To ensure the integrity of the request, the derived components `@method`
+and `@target-uri` SHOULD be verified. Additionally the integrity of the
+`content-digest` HTTP header and its validity SHOULD be checked.
+
+It SHOULD be verified that the `created` timestamp is within an
+acceptable range, allowing for time drift between servers.
+
+If this validation fails the response MUST use the HTTP status code
+`401` (Unauthorized).
```

Comment:
We could also add in usage of the `WWW-Authenticate` header to communicate error information back? Though Signature isn't an "Authorization" scheme. Perhaps using the [Problem Details for HTTP APIs](https://www.rfc-editor.org/rfc/rfc9457) proposed standard would be a good practice here?

Reply:
> We could also add in usage of the `WWW-Authenticate` header to communicate error information back? Though Signature isn't an "Authorization" scheme.

But `AWS4-HMAC-SHA256` seems to be, which is interesting because it is quite similar. So definitely something to watch out for as HTTP Message Signatures become more popular.

> Perhaps using the [Problem Details for HTTP APIs](https://www.rfc-editor.org/rfc/rfc9457) proposed standard would be a good practice here?

I would like to hold off on something like that until we have some real practical experience with the kind or problems that might occur here. So, not in v0.1 but maybe something for a v1.0.

Reply:
Maybe do it as an editors note: "In future versions we may use something like the Problem Details for HTTP APIs to distribute error information"